File: tests/refs/dutch_national_flag.cpp
// Reference C++ driver for tests/cases/dutch_national_flag.json.
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

void dutchFlag(std::vector<int>& a) {
    int lo = 0, mid = 0, hi = (int)a.size() - 1;
    while (mid <= hi) {
        if (a[mid] == 0) std::swap(a[lo++], a[mid++]);
        else if (a[mid] == 2) std::swap(a[mid], a[hi--]);
        else ++mid;
    }
}

int main() {
    std::string line, name;
    std::vector<int> arr, expected;
    bool hi = false, he = false; int p = 0, f = 0;
    int phase = 0; // 0=expecting input, 1=expecting expected

    auto run = [&]() {
        if (name.empty() || !hi || !he) return;
        auto got = arr; dutchFlag(got);
        if (got == expected) { std::cout << "PASS dutch_national_flag :: " << name << "\n"; ++p; }
        else {
            std::cout << "FAIL dutch_national_flag :: " << name << "\n  expected:";
            for (int v : expected) std::cout << ' ' << v;
            std::cout << "\n  got:";
            for (int v : got) std::cout << ' ' << v;
            std::cout << "\n"; ++f;
        }
        name.clear(); arr.clear(); expected.clear(); hi = he = false; phase = 0;
    };

    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line);
        std::string tag; iss >> tag;
        if (tag == "CASE") { iss >> name; phase = 0; }
        else if (tag == "ARR") {
            std::size_t n; iss >> n;
            std::vector<int> tmp(n);
            for (std::size_t i = 0; i < n; ++i) iss >> tmp[i];
            if (phase == 0) { arr = std::move(tmp); hi = true; phase = 1; }
            else { expected = std::move(tmp); he = true; }
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/dutch_national_flag.rs
// Reference Rust driver for tests/cases/dutch_national_flag.json.
use std::io::{self, BufRead};

fn dutch_flag(a: &mut Vec<i32>) {
    let (mut lo, mut mid) = (0usize, 0usize);
    if a.is_empty() { return; }
    let mut hi = a.len() - 1;
    loop {
        if mid > hi { break; }
        match a[mid] {
            0 => { a.swap(lo, mid); lo += 1; mid += 1; }
            2 => { a.swap(mid, hi); if hi == 0 { break; } hi -= 1; }
            _ => { mid += 1; }
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut arr: Vec<i32> = Vec::new();
    let mut expected: Vec<i32> = Vec::new();
    let mut hi = false; let mut he = false; let mut phase = 0;
    let mut p = 0u32; let mut f = 0u32;

    let flush = |name: &mut String, arr: &mut Vec<i32>, expected: &mut Vec<i32>,
                 hi: &mut bool, he: &mut bool, phase: &mut i32, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *hi && *he {
            let mut got = arr.clone();
            dutch_flag(&mut got);
            if got == *expected { println!("PASS dutch_national_flag :: {}", name); *p += 1; }
            else {
                println!("FAIL dutch_national_flag :: {}\n  expected: {:?}\n  got: {:?}", name, expected, got);
                *f += 1;
            }
        }
        name.clear(); arr.clear(); expected.clear(); *hi = false; *he = false; *phase = 0;
    };

    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() {
            flush(&mut name, &mut arr, &mut expected, &mut hi, &mut he, &mut phase, &mut p, &mut f);
            continue;
        }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { name = parts.next().unwrap_or("").to_string(); phase = 0; }
            "ARR" => {
                let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                let mut tmp: Vec<i32> = Vec::with_capacity(n);
                for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i32>() { tmp.push(v); } } }
                if phase == 0 { arr = tmp; hi = true; phase = 1; }
                else { expected = tmp; he = true; }
            }
            _ => {}
        }
    }
    flush(&mut name, &mut arr, &mut expected, &mut hi, &mut he, &mut phase, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
